src/cast_core/src/CollectionScanner.cpp

```diff
@@ -0,0 +1,102 @@
+// Copyright 2019-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/CollectionScanner.hpp>
+
+#include <memory>
+#include <chrono>
+
+#include <yaml-cpp/yaml.h>
+
+#include <bsoncxx/json.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+#include <mongocxx/database.hpp>
+
+#include <boost/log/trivial.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/context.hpp>
+
+#include <value_generators/DocumentGenerator.hpp>
+
+
+namespace genny::actor {
+
+struct CollectionScanner::PhaseConfig {
+    mongocxx::database database;
+    std::vector<std::string> collectionNames;
+    bool skipFirstLoop = false;
+    metrics::Operation scanOperation;
+    int actorId;
+    PhaseConfig(PhaseContext& context, const mongocxx::database& db, ActorId id, int collectionCount,
+      int threads, ActorCounter &counter)
+        : database{db},
+          skipFirstLoop{context["SkipFirstLoop"].maybe<bool>().value_or(false)},
+          scanOperation{context.operation("Scan", id)} {
+        actorId = counter;
+        counter ++;
+        // Distribute the collections among the actors.
+        collectionNames = CollectionScanner::getCollectionNames(collectionCount, threads, actorId);
+    }
+};
+
+void CollectionScanner::run() {
+    for (auto&& config : _loop) {
+        for (const auto&& _ : config) {
+            if (config->skipFirstLoop) {
+                config->skipFirstLoop = false;
+            } else {
+                try {
+                    _runningActorCounter ++;
+                    BOOST_LOG_TRIVIAL(info) << "Starting collection scanner id:" << config->actorId;
+                    // Count over all collections this thread has been tasked with scanning each.
+                    auto statTracker = config->scanOperation.start();
+                    for (int i = 0; i < config->collectionNames.size(); ++i){
+                        statTracker.addDocuments(config->database[config->collectionNames[i]].count_documents({}));
```

Comment:
1. You can use range-based loops here

    ```cpp
    for (const auto&& collectionName : this->config->collectionNames)`
    ```

2. You may also consider having a `std::vector<collection>` rather than `std::vector<string>` and doing the `database[collectionName]` operation at Actor-setup time. This would save the time it takes to allocate and deallocate the `mongocxx::collection` type. Not a huge deal but probably an easy change nonetheless.

Reply:
✔